### DISCUSSION TITLE
Keyboard and mouse sharing tools... Unite!

### DISCUSSION THREAD

**Discussion by nbolton** [deskflow/deskflow#7590] [Development]
Recently, @sithlord48 and I have been discussing the idea of a library called CISP or `libcisp` (common input sharing parts). 

Our goal is to unite Deskflow, Input Leap, Barrier*, Synergy, and Lan Mouse (@feschber) in terms of interoperability.

\* Barrier is an odd case as it's seemingly a dead project, but many people are still using it who haven't switched to Input Leap or Deskflow.

The first stage of CISP will be to create a common protocol with a common handshake which can be used by all open source keyboard and mouse sharing projects, to enable interoperability.

Next step... we don't want to get too far ahead of ourselves, but we hope to share as much as possible beyond just the protocol. But the protocol will be a good starting point.

Also, the name is open to discussion (`libflow` was another idea), and of course, anything else is also up for discussion. Suggestions welcome as always!

**Comment by sithlord48** [comment]
There is a [repo](https://github.com/sithlord48/cisp) that i started but there is not much here yet just a skeleton.

**Comment by feschber** [comment]
Hi @nbolton, hi @sithlord48, I really like the idea of interoperability and am glad to hear you are actively working toward this!

I have some thoughts on this:
1. Maybe it would be a good idea to define the protocol in a language agnostic way through e.g. xml as is done in [libei](https://gitlab.freedesktop.org/libinput/libei/-/blob/main/proto/protocol.xml?ref_type=heads) or wayland. This would enable native language bindings, which I would prefer to depending on a library.
2. I dont know what the plan is, but I think it would be neat to separate transport and application logic here to enable different transport protocols. Input-leap is using tcp, lan-mouse is using udp / dtls and I was thinking of exploring Quic as well in the future.
3. I'm not too familiar with synergy / deskflow and have not taken a deeper look at the networkprotocol before but to my knowledge, you are exchanging keycodes instead of raw scancodes. This would need to be considered, as lan-mouse is using raw scancodes.
4. Similar point to (3) is pointer motion events, which could be either "client-side accelerated" or "server-side accelerated". I've had a bit of a disscussion about this [over here](https://github.com/feschber/lan-mouse/issues/156). Again I do not know what synergy is doing, but it would need to be a consideration for the protocol.

In any case, you can have a look at the lan-mouse protocol [over here](https://github.com/feschber/lan-mouse/tree/main/lan-mouse-proto).
I've changed a few things in the development branch but the general idea is the same. The modifier event is likely going to get dropped in a future version and the Enter event may also carry additional information about where a device was entered at some point.

**Reply by sithlord48** [reply]
Hello and thank you for your reply. 
we have an issue open for a protocol description (https://github.com/input-leap/input-leap/issues/1964). 

I was thinking since both input-leap and deskflow(synergy) are built with C++ A good place to start would be to take what we have that is in our "libs" for both projects and create a shared lib they could both use . Making sure they are always compatible  and having a central place for the development of a drop in usable library.  I had figured that it would be easier to work to get this into those two projects first.  This would be our initial unification step. 

My goal here is two fold. 

1. I want to ensure compatibility across the various projects doing this similar thing . By Having a common set of basic features each app could use to really do what they want with the input sharing . This ideally should make it trivial to make a server / client with `libcisp` . We would also not only cover the protocol but ideally the basic platform parts as well ( input capture , emulation and clipboard support ) .  Why the P is for Parts not protocol

2. Provide an easy way to "swap" these parts so that it would be easier to rewrite these parts using other tech.  I suspect that most of what we are writing that is platform specific can be done with Qt and would like to explore this possibility, There has been talk about making the parts in rust and having this split here also allows us to try this also easier without having to throw out the entire working application.
